bacula_server_operator/src/charm.py

```diff
@@ -0,0 +1,326 @@
+#!/usr/bin/env python3
+
+# Copyright 2025 Canonical Ltd.
+# See LICENSE file for licensing details.
+
+# Learn more at: https://documentation.ubuntu.com/juju/3.6/howto/manage-charms/#build-a-charm
+
+"""Bacula-server charm the service."""
+
+import json
+import logging
+import secrets
+import typing
+import urllib.parse
+
+import ops
+
+import charms.data_platform_libs.v0.data_interfaces as data_interfaces
+import charms.data_platform_libs.v0.s3 as s3
+from . import bacula
+from . import relations
+
+PEER_RELATION_NAME = "bacula-peer"
+BACULUM_CHARM_USERNAME = "charm-admin"
+
+logger = logging.getLogger(__name__)
+
+
+class NotReadyError(Exception):
+    pass
+
+
+class UnrecoverableError(Exception):
+    pass
+
+
+class BaculaServerCharm(ops.CharmBase):
+    """Charm the service."""
+
+    def __init__(self, *args: typing.Any):
+        """Construct.
+
+        Args:
+            args: Arguments passed to the CharmBase parent constructor.
+        """
+        super().__init__(*args)
+        self._database = data_interfaces.DatabaseRequires(
+            charm=self,
+            relation_name="postgresql",
+            database_name=self.app.name,
+        )
+        self._s3 = s3.S3Requirer(
+            charm=self,
+            relation_name="s3",
+            bucket_name=self.app.name,
+        )
+        self._bacula = bacula.Bacula()
+        self._bacula_dir_relation = relations.BaculaProvider(charm=self)
+
+        self.framework.observe(self.on.postgresql_relation_changed, self._reconcile_event)
+        self.framework.observe(self.on.postgresql_relation_broken, self._reconcile_event)
+        self.framework.observe(self.on.s3_relation_changed, self._reconcile_event)
+        self.framework.observe(self.on.s3_relation_broken, self._reconcile_event)
+        self.framework.observe(self.on.config_changed, self._reconcile_event)
+        self.framework.observe(self.on.upgrade_charm, self._reconcile_event)
+        self.framework.observe(self.on.secret_changed, self._reconcile_event)
```

Comment:
We should listen to the secret removed hook

Reply:
What the reason behind listen to the `secret-removed` hook?

Reply:
To avoid piling up secret revisions and probably remove the secret when the relation is broken